content/2020-04-27-rendering-a-list-in-graphql-and-reasonreact/index.md

```diff
@@ -0,0 +1,168 @@
+
+Following their `README`, I set everything up and it looked good! 
+
+I was even able to log my data to the console with `Js.log2("my data", data)` and it worked! 
+
+If only rendering data was as easy as logging in...
+
+### What I struggled with
+
+When it came to rendering data, I struggled. The GraphQL API I was working with had a lot of nullable data. In regular vanilla React, I don't really worry about this kind of thing. Or at least, I don't think I do.
+
+However, because Reason is typed, I do have to worry about it. And this is where I got stuck. 
+
+I was stuck on this particular error message:
+
+```shell
+Error: This pattern matches values of type option('a)
+       but a pattern was expected which matches values of type
+         Js.Array.t(option({. "currency": option(string),
+                             "rate": option(string)}))
+           =
+           array(option({. "currency": option(string),
+                          "rate": option(string)}))
+```
+
+I didn't understand the option part. It looked funky to me. After trying to google various things and struggling for a few hours, I finally decided to post in the [Reason Discord](https://discordapp.com/invite/reasonml).
+
+I hadn't spent much time in the Reason Discord community prior to this so I didn't know what to expect.
+
+To my surprise, people were quick to help me! At first, I didn't understand the solutions people gave me. I felt embarrassed to ask more questions and have them clarify, but after I did, it cleared things up.
+
+With a little bit of extra help for rendering an array of `null` (due to the way the API was designed + the types Reason wanted), I figured out the solution!
+
+![Some data or none, drawing](drawing-data.png)
+<small>
+    Open this drawing in <a href="https://excalidraw.com/#json=4900521227845632,zu4fe5MiNz4AHPb76JnA6A" target="_blank">Excalidraw</a> 
+</small>
+
+## Solution
+
+Here's what the code looked like after getting things to work:
+
+```reason
+open ApolloHooks;
+open Belt;
+
+module GetExchangeRates = [%graphql
+  {|
+  query getExchangeRates {
+    rates(currency: "USD") {
+      currency
+      rate
+    }
+  }
+|}
+];
+
+[@react.component]
+let make = () => {
+  let (simple, _full) = useQuery(GetExchangeRates.definition);
+
+  <div>
+    {switch (simple) {
+     | Loading => <p> {React.string("Loading...")} </p>
+     | Data(data) =>
+        let currencies =
+          switch (data##rates) {
+          | Some(rates) =>
+            rates->Array.map(rate =>
+              rate
+              ->Option.map(rate => rate##currency)
+              ->Option.mapWithDefault(React.null, currency =>
+                  switch (currency) {
+                  | Some(currency) => <p> {React.string(currency)} </p>
+                  | None => React.null
```

Comment:
Thanks so much for the suggestion @fhammerschmidt - I wasn't familiar with Option, but definitely makes it nicer/easier to read! I had to switch on the last statement because it was an option string, but expected a string. This made it work though!

Reply:
Ah, so if currency is also an option, you can do 
`Option.flatMap(rate => rate##currency)` instead of `Option.map` and get rid of the inner switch that way.

Reply:
Ah! I would have _never_ figured that out on my own. I'm going to update the sample app and this example. Thanks a ton!